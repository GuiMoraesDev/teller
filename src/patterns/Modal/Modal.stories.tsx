import React from 'react';

import { ComponentMeta, ComponentStory } from '@storybook/react';

import Modal from '.';

const Storie = {
	title: 'Patterns/Modal',
	component: Modal,
} as ComponentMeta<typeof Modal>;

export default Storie;

const Template: ComponentStory<typeof Modal> = (args) => (
	<Modal {...args}>Your content here</Modal>
);

export const Default = Template.bind({});

Default.args = {};
