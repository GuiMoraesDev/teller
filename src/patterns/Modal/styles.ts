import styled from 'styled-components';

interface ModalContainerProps {
	isVisible: boolean;
}

export const ModalContainer = styled.div<ModalContainerProps>`
	position: absolute;

	top: 0;
	bottom: 0;
	right: 0;
	left: 0;

	display: flex;

	align-items: center;
	justify-content: center;

	width: 100vw;
	height: 100vh;

	overflow: hidden;

	background-color: rgba(0, 0, 0, 0.6);

	transition: all ${({ theme }) => theme.transition.fast};

	opacity: ${({ isVisible }) => (isVisible ? 1 : 0)};
	visibility: ${({ isVisible }) => (isVisible ? 'visible' : 'hidden')};
`;

export const ModalContentWrapper = styled.main`
	position: relative;

	display: grid;

	grid-template-rows: ${({ theme }) => theme.sizes.dimensions.height.xs} auto;

	min-width: ${({ theme }) => theme.sizes.common.x20};
	min-height: ${({ theme }) => theme.sizes.common.x20};

	background-color: ${({ theme }) => theme.themeColors.canvas};

	border-radius: ${({ theme }) => theme.rounded.sm};
`;

export const Header = styled.header`
	display: flex;

	align-items: center;
	justify-content: flex-end;
`;

export const ModalContent = styled.section`
	padding: ${({ theme }) => theme.sizes.dimensions.spacing.xs};
`;
