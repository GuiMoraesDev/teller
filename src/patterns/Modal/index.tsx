import React, { forwardRef, useImperativeHandle, useState } from 'react';
import { PropsWithChildren, useCallback } from 'react';

import { X } from 'phosphor-react';

import Button from 'components/Button';

import * as Styles from './styles';

export interface ModalHandles {
	openModal: () => void;
	closeModal: () => void;
}

interface ModalProps {
	startOpened?: boolean;
}

const Modal: React.ForwardRefRenderFunction<
	ModalHandles,
	PropsWithChildren<ModalProps>
> = ({ children, startOpened }, ref): JSX.Element => {
	const isStorybook = React.useMemo(() => process.env.IS_STORYBOOK, []);

	const [visible, setVisible] = useState(!!isStorybook || !!startOpened);

	const openModal = useCallback(() => {
		setVisible(true);
	}, []);

	const closeModal = useCallback(() => {
		setVisible(false);
	}, []);

	useImperativeHandle(ref, () => ({
		openModal,
		closeModal,
	}));

	return (
		<Styles.ModalContainer isVisible={visible} data-testid="modal-container">
			<Styles.ModalContentWrapper>
				<Styles.Header>
					<Button
						IconLeft={<X />}
						onClick={closeModal}
						variant="neutral"
						dimension="xs"
						data-testid="close-button"
					/>
				</Styles.Header>

				<Styles.ModalContent data-testid="modal-content">
					{children}
				</Styles.ModalContent>
			</Styles.ModalContentWrapper>
		</Styles.ModalContainer>
	);
};

export default forwardRef(Modal);
