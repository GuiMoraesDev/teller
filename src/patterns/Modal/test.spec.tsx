import React from 'react';

import userEvent from '@testing-library/user-event';

import { render, waitFor } from 'tests/test.utils';

import Modal from './index';

const ContainerRender = (startOpened?: boolean) =>
	render(<Modal startOpened={startOpened} />);

const elementsTestId = {
	modalContainer: 'modal-container',
	closeButton: 'close-button',
	modalContent: 'modal-content',
};

describe('Modal pattern', () => {
	it('should render correctly', () => {
		const { container, getByTestId } = ContainerRender();

		const modalContainer = getByTestId(elementsTestId.modalContainer);
		const closeButton = getByTestId(elementsTestId.closeButton);
		const modalContent = getByTestId(elementsTestId.modalContent);

		expect(modalContainer).not.toBeVisible();
		expect(closeButton).not.toBeVisible();
		expect(modalContent).not.toBeVisible();

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render Modal opened without user interaction', () => {
		const { getByTestId } = ContainerRender(true);

		const modalContainer = getByTestId(elementsTestId.modalContainer);
		const closeButton = getByTestId(elementsTestId.closeButton);
		const modalContent = getByTestId(elementsTestId.modalContent);

		expect(modalContainer).toBeVisible();
		expect(closeButton).toBeVisible();
		expect(modalContent).toBeVisible();
	});

	it('should close Modal on close button click', async () => {
		const { getByTestId } = ContainerRender(true);

		const closeButton = getByTestId(elementsTestId.closeButton);
		const modalContent = getByTestId(elementsTestId.modalContent);

		await waitFor(() => userEvent.click(closeButton));

		expect(modalContent).not.toBeVisible();
	});
});
