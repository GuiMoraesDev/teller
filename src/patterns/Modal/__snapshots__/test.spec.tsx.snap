// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal pattern should render correctly 1`] = `
.c3 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 3.6rem;
  border: 0.1rem solid #C9CAC8;
  border-radius: 0.4rem;
  margin: 0;
  padding: 0 0.8rem;
  gap: 1.2rem;
  text-transform: none;
  -webkit-transition: background-color 0.2s, color 0.2s;
  transition: background-color 0.2s, color 0.2s;
  cursor: pointer;
  color: #181917;
  background-color: transparent;
  border-color: transparent;
}

.c3:hover {
  background-color: #82ADB2;
}

.c3 svg {
  color: inherit;
}

.c3:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #181917;
  background-color: transparent;
}

.c3 .spin {
  -webkit-animation-name: ieAXeN;
  animation-name: ieAXeN;
  -webkit-animation-duration: 2s;
  animation-duration: 2s;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
}

.c0 {
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100vw;
  height: 100vh;
  overflow: hidden;
  background-color: rgba(0,0,0,0.6);
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  opacity: 0;
  visibility: hidden;
}

.c1 {
  position: relative;
  display: grid;
  grid-template-rows: 3.6rem auto;
  min-width: 8.0rem;
  min-height: 8.0rem;
  background-color: #FDFDFD;
  border-radius: 0.4rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c4 {
  padding: 0.8rem;
}

<div
  class="c0"
  data-testid="modal-container"
>
  <main
    class="c1"
  >
    <header
      class="c2"
    >
      <button
        class="c3"
        data-testid="close-button"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="none"
            height="256"
            width="256"
          />
          <line
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
            x1="200"
            x2="56"
            y1="56"
            y2="200"
          />
          <line
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
            x1="200"
            x2="56"
            y1="200"
            y2="56"
          />
        </svg>
      </button>
    </header>
    <section
      class="c4"
      data-testid="modal-content"
    />
  </main>
</div>
`;
